File: packages/governance-ts/examples/basic-governance.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

/**
 * basic-governance.ts
 *
 * Demonstrates the core @aumos/governance workflow:
 *   1. Configure and instantiate GovernanceEngine.
 *   2. Manually assign trust levels to agents.
 *   3. Create spending budgets per action category.
 *   4. Evaluate actions and inspect decisions.
 *   5. Query the audit log for a historical view.
 *
 * Run: npx tsx examples/basic-governance.ts
 */

import {
  GovernanceEngine,
  TrustLevel,
} from '../src/index.js';

async function main(): Promise<void> {
  // --------------------------------------------------------------------------
  // 1. Initialise GovernanceEngine with inline configuration.
  // --------------------------------------------------------------------------
  const engine = new GovernanceEngine({
    trust: {
      defaultLevel: TrustLevel.L0_OBSERVER,
    },
    budget: {
      envelopes: [
        { category: 'communication', limit: 10.00, period: 'daily' },
        { category: 'external_api',  limit: 25.00, period: 'daily' },
        { category: 'data_access',   limit: 5.00,  period: 'hourly' },
      ],
    },
    consent: {
      requireConsent: true,
      defaultPurposes: ['audit', 'monitoring'],
    },
    audit: {
      enabled: true,
      maxRecords: 500,
    },
  });

  const agentId = 'agent:demo-assistant-v1';

  // --------------------------------------------------------------------------
  // 2. Assign a trust level to the agent.
  //    Trust is MANUAL ONLY — there is no auto-promotion pathway.
  // --------------------------------------------------------------------------
  const assignment = engine.trust.setLevel(
    agentId,
    TrustLevel.L3_ACT_APPROVE,
    undefined,
    { reason: 'Approved by platform admin after review.' },
  );
  console.log(
    `Trust assigned: ${assignment.level} (${TrustLevel[assignment.level]}) ` +
    `to ${assignment.agentId}`,
  );

  // --------------------------------------------------------------------------
  // 3. Evaluate a communication action that should be permitted.
  // --------------------------------------------------------------------------
  const emailDecision = await engine.evaluate({
    agentId,
    action: 'send_email',
    category: 'communication',
    requiredTrustLevel: TrustLevel.L3_ACT_APPROVE,
    cost: 0.002,
  });

  console.log('\n--- send_email ---');
  console.log('Permitted:', emailDecision.permitted);
  console.log('Reason:   ', emailDecision.reason);
  console.log('Protocol: ', emailDecision.protocol);

  // --------------------------------------------------------------------------
  // 4. Evaluate an action that requires consent for PII data access.
  //    No consent has been recorded yet — expect denial.
  // --------------------------------------------------------------------------
  const piiDenied = await engine.evaluate({
    agentId,
    action: 'read_customer_profile',
    category: 'data_access',
    requiredTrustLevel: TrustLevel.L2_SUGGEST,
    cost: 0.001,
    dataType: 'pii',
    purpose: 'personalisation',
  });

  console.log('\n--- read_customer_profile (no consent) ---');
  console.log('Permitted:', piiDenied.permitted);
  console.log('Reason:   ', piiDenied.reason);

  // --------------------------------------------------------------------------
  // 5. Record consent, then re-evaluate the same action — expect permit.
  // --------------------------------------------------------------------------
  engine.consent.recordConsent(
    agentId,
    'pii',
    'personalisation',
    'operator:alice',
  );

  const piiPermitted = await engine.evaluate({
    agentId,
    action: 'read_customer_profile',
    category: 'data_access',
    requiredTrustLevel: TrustLevel.L2_SUGGEST,
    cost: 0.001,
    dataType: 'pii',
    purpose: 'personalisation',
  });

  console.log('\n--- read_customer_profile (consent granted) ---');
  console.log('Permitted:', piiPermitted.permitted);
  console.log('Reason:   ', piiPermitted.reason);

  // --------------------------------------------------------------------------
  // 6. Attempt an action that exceeds the trust requirement.
  // --------------------------------------------------------------------------
  const autonomousDenied = await engine.evaluate({
    agentId,
    action: 'deploy_to_production',
    category: 'system',
    requiredTrustLevel: TrustLevel.L5_AUTONOMOUS,
    cost: 0,
  });

  console.log('\n--- deploy_to_production (trust too low) ---');
  console.log('Permitted:', autonomousDenied.permitted);
  console.log('Reason:   ', autonomousDenied.reason);

  // --------------------------------------------------------------------------
  // 7. Inspect the audit log.
  // --------------------------------------------------------------------------
  const allRecords = engine.audit.getRecords();
  console.log(`\nAudit log contains ${allRecords.length} record(s).`);

  const denials = engine.audit.query({ outcome: 'deny' });
  console.log(`Denied decisions: ${denials.length}`);

  const permits = engine.audit.query({ outcome: 'permit' });
  console.log(`Permitted decisions: ${permits.length}`);

  // --------------------------------------------------------------------------
  // 8. Budget utilisation snapshot.
  // --------------------------------------------------------------------------
  const communicationUtilization = engine.budget.getUtilization('communication');
  if (communicationUtilization !== undefined) {
    console.log(
      `\nCommunication budget: $${communicationUtilization.spent.toFixed(4)} / ` +
      `$${communicationUtilization.limit.toFixed(2)} ` +
      `(${communicationUtilization.utilizationPercent.toFixed(1)}% used)`,
    );
  }
}

main().catch((error: unknown) => {
  console.error('Example failed:', error);
  process.exit(1);
});

File: packages/governance-py/examples/basic_governance.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation
"""
Basic governance example.

Demonstrates creating a GovernanceEngine and running it through the
trust, budget, and consent checks for a simple agent action.

Run with:
    python examples/basic_governance.py
"""
from __future__ import annotations

import asyncio

from aumos_governance import (
    AuditFilter,
    GovernanceAction,
    GovernanceConfig,
    GovernanceEngine,
    GovernanceOutcome,
    TrustLevel,
    aggregate_outcomes,
)
from aumos_governance.config import AuditConfig, BudgetConfig, ConsentConfig, TrustConfig


async def main() -> None:
    # ------------------------------------------------------------------ #
    # 1. Configure and build the engine
    # ------------------------------------------------------------------ #
    config = GovernanceConfig(
        trust=TrustConfig(default_level=1, enable_decay=False),
        budget=BudgetConfig(allow_overdraft=False, rollover_on_reset=False),
        consent=ConsentConfig(default_deny=True),
        audit=AuditConfig(max_records=500, include_context=True),
    )
    engine = GovernanceEngine(config=config)

    # ------------------------------------------------------------------ #
    # 2. Set up trust, budget, and consent records
    # ------------------------------------------------------------------ #
    engine.trust.set_level("agent-alpha", TrustLevel.L3_ACT_APPROVE)
    engine.trust.set_level("agent-beta", TrustLevel.L1_MONITOR)

    engine.budget.create_budget("llm-calls", limit=100.0, period="monthly")
    engine.budget.create_budget("tool-executions", limit=50.0, period="daily")

    engine.consent.record_consent(
        agent_id="agent-alpha",
        data_type="user_profile",
        purpose="personalisation",
        granted_by="admin@example.com",
    )

    # ------------------------------------------------------------------ #
    # 3. Evaluate actions
    # ------------------------------------------------------------------ #
    print("=== Example 1: Approved action ===")
    decision = await engine.evaluate(
        GovernanceAction(
            agent_id="agent-alpha",
            required_trust_level=TrustLevel.L2_SUGGEST,
            budget_category="llm-calls",
            budget_amount=2.5,
            data_type="user_profile",
            purpose="personalisation",
            action_type="llm_completion",
            resource="gpt-4o",
        )
    )
    print(f"  allowed: {decision.allowed}")
    print(f"  outcome: {decision.outcome}")
    for reason in decision.reasons:
        print(f"  - {reason}")

    print()
    print("=== Example 2: Denied — trust too low ===")
    decision2 = await engine.evaluate(
        GovernanceAction(
            agent_id="agent-beta",
            required_trust_level=TrustLevel.L3_ACT_APPROVE,
            action_type="file_write",
            resource="/etc/config",
        )
    )
    print(f"  allowed: {decision2.allowed}")
    print(f"  outcome: {decision2.outcome}")
    for reason in decision2.reasons:
        print(f"  - {reason}")

    print()
    print("=== Example 3: Denied — consent not granted ===")
    decision3 = await engine.evaluate(
        GovernanceAction(
            agent_id="agent-alpha",
            data_type="financial_records",
            purpose="reporting",
            action_type="data_read",
        )
    )
    print(f"  allowed: {decision3.allowed}")
    print(f"  outcome: {decision3.outcome}")
    for reason in decision3.reasons:
        print(f"  - {reason}")

    # ------------------------------------------------------------------ #
    # 4. Query the audit log
    # ------------------------------------------------------------------ #
    print()
    print("=== Audit log summary ===")
    all_results = engine.audit.query()
    stats = aggregate_outcomes(all_results.records)
    print(f"  total: {stats['total']}")
    print(f"  allow: {stats['allow']}")
    print(f"  deny:  {stats['deny']}")
    print(f"  denial_rate: {stats['denial_rate']:.0%}")

    denied_results = engine.audit.query(AuditFilter(outcome=GovernanceOutcome.DENY))
    print(f"\n  Denied records ({denied_results.total_matched}):")
    for record in denied_results.records:
        agent = record.context.agent_id if record.context else "unknown"
        print(f"    [{record.record_id[:8]}] agent={agent} — {record.decision}")

    # ------------------------------------------------------------------ #
    # 5. Demonstrate evaluate_sync
    # ------------------------------------------------------------------ #
    print()
    print("=== Example 4: Synchronous evaluation ===")
    sync_decision = engine.evaluate_sync(
        GovernanceAction(
            agent_id="agent-alpha",
            required_trust_level=TrustLevel.L2_SUGGEST,
            action_type="health_check",
        )
    )
    print(f"  allowed: {sync_decision.allowed}")
    print(f"  audit_record_id: {sync_decision.audit_record_id}")


if __name__ == "__main__":
    asyncio.run(main())
